File: src/bmsx_cpp/machine/devices/device_status.cpp
#include "machine/devices/device_status.h"

namespace bmsx {

DeviceStatusLatch::DeviceStatusLatch(Memory& memory, DeviceStatusRegisters registers)
	: code(registers.noneCode)
	, m_memory(memory)
	, m_registers(registers) {
	detail = 0u;
}

void DeviceStatusLatch::resetStatus() const {
	status = 0u;
	code = m_registers.noneCode;
	detail = 0u;
	writeRegisterState();
}

void DeviceStatusLatch::restore(uint32_t restoredStatus, uint32_t restoredCode, uint32_t restoredDetail) const {
	status = restoredStatus;
	code = restoredCode;
	detail = restoredDetail;
	writeRegisterState();
}

void DeviceStatusLatch::writeRegisterState() const {
	m_memory.writeIoValue(m_registers.statusAddr, valueNumber(static_cast<double>(status)));
	m_memory.writeIoValue(m_registers.codeAddr, valueNumber(static_cast<double>(code)));
	m_memory.writeIoValue(m_registers.detailAddr, valueNumber(static_cast<double>(detail)));
	m_memory.writeIoValue(m_registers.ackAddr, valueNumber(0.0));
}

void DeviceStatusLatch::clear() const {
	code = m_registers.noneCode;
	detail = 0u;
	m_memory.writeIoValue(m_registers.codeAddr, valueNumber(static_cast<double>(code)));
	m_memory.writeIoValue(m_registers.detailAddr, valueNumber(static_cast<double>(detail)));
	setStatusFlag(m_registers.faultMask, false);
}

void DeviceStatusLatch::acknowledge() const {
	if (m_memory.readIoU32(m_registers.ackAddr) == 0u) {
		return;
	}
	clear();
	m_memory.writeIoValue(m_registers.ackAddr, valueNumber(0.0));
}

void DeviceStatusLatch::setStatusFlag(uint32_t mask, bool active) const {
	const uint32_t nextStatus = active ? (status | mask) : (status & ~mask);
	if (nextStatus == status) {
		return;
	}
	status = nextStatus;
	m_memory.writeIoValue(m_registers.statusAddr, valueNumber(static_cast<double>(status)));
}

void DeviceStatusLatch::raise(uint32_t raisedCode, uint32_t raisedDetail) const {
	if ((status & m_registers.faultMask) != 0u) {
		return;
	}
	code = raisedCode;
	detail = raisedDetail;
	m_memory.writeIoValue(m_registers.codeAddr, valueNumber(static_cast<double>(code)));
	m_memory.writeIoValue(m_registers.detailAddr, valueNumber(static_cast<double>(detail)));
	setStatusFlag(m_registers.faultMask, true);
}

} // namespace bmsx

File: packages/bmsx-console/src/machine/devices/device_status.ts
import type { Memory } from '../memory/memory';

export type DeviceStatusRegisters = {
	statusAddr: number;
	codeAddr: number;
	detailAddr: number;
	ackAddr: number;
	faultMask: number;
	noneCode: number;
};

export class DeviceStatusLatch {
	public status = 0;
	public code: number;
	public detail = 0;

	public constructor(
		private readonly memory: Memory,
		private readonly registers: DeviceStatusRegisters,
	) {
		this.code = registers.noneCode;
		this.detail = 0;
	}

	public resetStatus(): void {
		this.status = 0;
		this.code = this.registers.noneCode;
		this.detail = 0;
		this.writeRegisterState();
	}

	public restore(status: number, code: number, detail: number): void {
		this.status = status >>> 0;
		this.code = code >>> 0;
		this.detail = detail >>> 0;
		this.writeRegisterState();
	}

	private writeRegisterState(): void {
		this.memory.writeIoValue(this.registers.statusAddr, this.status);
		this.memory.writeIoValue(this.registers.codeAddr, this.code);
		this.memory.writeIoValue(this.registers.detailAddr, this.detail);
		this.memory.writeIoValue(this.registers.ackAddr, 0);
	}

	public clear(): void {
		this.code = this.registers.noneCode;
		this.detail = 0;
		this.memory.writeIoValue(this.registers.codeAddr, this.code);
		this.memory.writeIoValue(this.registers.detailAddr, this.detail);
		this.setStatusFlag(this.registers.faultMask, false);
	}

	public acknowledge(): void {
		if (this.memory.readIoU32(this.registers.ackAddr) === 0) {
			return;
		}
		this.clear();
		this.memory.writeIoValue(this.registers.ackAddr, 0);
	}

	public setStatusFlag(mask: number, active: boolean): void {
		const nextStatus = active ? (this.status | mask) : (this.status & ~mask);
		if (nextStatus === this.status) {
			return;
		}
		this.status = nextStatus >>> 0;
		this.memory.writeIoValue(this.registers.statusAddr, this.status);
	}

	public raise(code: number, detail: number): void {
		if ((this.status & this.registers.faultMask) !== 0) {
			return;
		}
		this.code = code >>> 0;
		this.detail = detail >>> 0;
		this.memory.writeIoValue(this.registers.codeAddr, this.code);
		this.memory.writeIoValue(this.registers.detailAddr, this.detail);
		this.setStatusFlag(this.registers.faultMask, true);
	}
}
